### PR TITLE
[4.0] [a11y] [NO CACHE] Make media manager browse view work with keyboard navigation

### PR DESCRIPTION
**Pull request by wilsonge**
This fixes the browse view aspect of what's discussed in https://github.com/joomla/joomla-cms/issues/23932 (please do not close that issue as the list view is still to be worked on)

Fixes various a11y issue keyboard navigation issues in the media manager browse view

### CODE REVIEW

**Comment by zwiastunsw** [comment]
Link "Toggle select all" should be a checkbox. Anachor is an incorrect tag at this point. It is not possible to communicate the status of the switch to the users of the screen reader.
mm_toolbar_toggle png_

```html
    <div class="media-view-icons">
      <a href="#" aria-label="Toggle select all" class="media-toolbar-icon media-toolbar-select-all">
        <span aria-hidden="true" class="fa fa-square-o"></span>
      </a>
    </div> 
```

**Comment by zwiastunsw** [comment]
The part of the toolbar is the breadcrumbs. It should be included in the `nav` tag with `aria-label="You are here"` (e.g)
mm_breadcrumbs

```
    <ol class="media-breadcrumb">
      <li class="media-breadcrumb-item">
        <a href="#">images</a>
      </li>
      <li class="media-breadcrumb-item">
        <a href="#">banners</a>
      </li>
    </ol> 
```

**Comment by zwiastunsw** [comment]
The toolbar should have the attribute role="toolbar" and the attribute aria-label="Media manager" (e.g). 

mm_toolbar

```html
  <div class="media-toolbar" role="toolbar" aria-label="COM_MEDIA_TOOLBAR">
    <!----> 
  </div>

**Comment by zwiastunsw** [comment]
In the toolbar we have a set of buttons. All of them are incorrectly tagged as links.
They should indicate to the user the status - `a` tag a cannot be used to indicate a state.

In addition: Grid switches (increase, decrease) should be hidden from screen readers.

mm_toolbar_buttons

**Comment by wilsonge** [comment]
Ok the toolbar role/label change is easy to do. What’s the recommended markup for the breadcrumbs? Should be easy enough

I’m leaving the checkboxes for another PR. Implementing a form element is going to be complicated and is not something I want to try doing as part of this PR

**Comment by brianteeman** [comment]
Look at the breadcrumbs module o think that's correct

**Comment by zwiastunsw** [comment]
```html
<nav role="navigation" aria-label="COM_MEDIA_BREADCRUMBS">
       <ol class="media-breadcrumb">
         <li class="media-breadcrumb-item">
           <a href="#">images</a>
         </li>
       </ol>
   </nav>
```

**Comment by wilsonge** [comment]
OK Fixes to toolbar - it now has a `role=toolbar`, aria-label, and all the buttons are now `<button>` elements instead of `<a>` tags. Breadcrumbs fixes also pushed up. I deliberately didn't add the `role=nav` in the example @zwiastunsw gave because https://www.w3.org/TR/wai-aria-practices/examples/breadcrumb/index.html was missing it and added in the `aria-current` mentioned there. But happy to flip back if needed.

**Comment by wilsonge** [comment]
I think that's everything covered aside from the checkboxes you guys have asked for

File: administrator/components/com_media/resources/scripts/components/browser/items/file.vue [hunk 1]
```diff
@@ -12,31 +12,36 @@
...
         <a href="#" class="media-browser-select"
           @click.stop="toggleSelect()"
-          :aria-label="translate('COM_MEDIA_TOGGLE_SELECT_ITEM')">
+          :aria-label="translate('COM_MEDIA_TOGGLE_SELECT_ITEM')"
+          @focus="focused(true)" @blur="focused(false)">
         </a>
         <div class="media-browser-actions" :class="{'active': showActions}">
-            <a href="#" class="action-toggle"
-              :aria-label="translate('COM_MEDIA_OPEN_ITEM_ACTIONS')">
+            <a href="#" class="action-toggle" role="button"
```
Comment (brianteeman) [inline administrator/components/com_media/resources/scripts/components/browser/items/file.vue]: If it looks like a button and acts like a button then it should be a button

File: administrator/components/com_media/resources/scripts/components/browser/items/image.vue [hunk 2]
```diff
@@ -10,41 +10,49 @@
...
         <a href="#" class="media-browser-select"
           @click.stop="toggleSelect()"
-          :aria-label="translate('COM_MEDIA_TOGGLE_SELECT_ITEM')">
+          :aria-label="translate('COM_MEDIA_TOGGLE_SELECT_ITEM')"
+          @focus="focused(true)" @blur="focused(false)">
         </a>
         <div class="media-browser-actions" :class="{'active': showActions}">
-            <a href="#" class="action-toggle"
-              :aria-label="translate('COM_MEDIA_OPEN_ITEM_ACTIONS')">
+            <a href="#" class="action-toggle" role="button"
```
Comment (brianteeman) [inline administrator/components/com_media/resources/scripts/components/browser/items/image.vue]: If it looks like a button and acts like a button then it should be a button

File: administrator/components/com_media/resources/scripts/components/toolbar/toolbar.vue [hunk 3]
```diff
@@ -13,30 +13,32 @@
             <input type="text" @input="changeSearch" :placeholder="translate('COM_MEDIA_SEARCH')"/>
         </div>
         <div class="media-view-icons">
-            <a href="#" class="media-toolbar-icon media-toolbar-decrease-grid-size"
+            <button role="button" class="media-toolbar-icon media-toolbar-decrease-grid-size"
```
Comment (brianteeman) [inline administrator/components/com_media/resources/scripts/components/toolbar/toolbar.vue]: these should be type=button not role=button

File: administrator/components/com_media/tmpl/media/default_texts.php [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
 use Joomla\CMS\Language\Text;
 
-Text::script('COM_MEDIA_ACTIN_RENAME', true);
+Text::script('COM_MEDIA_ACTION_RENAME', true);
```
Comment (brianteeman) [inline administrator/components/com_media/tmpl/media/default_texts.php:13]: I guess someone is going to ask for this to be alpha ordered

File: administrator/language/en-GB/en-GB.com_media.ini [hunk 5]
```diff
@@ -4,11 +4,12 @@
 ; Note : All ini files need to be saved as UTF-8
 
 COM_MEDIA="Media"
-COM_MEDIA_ACTIN_RENAME="Rename item"
+COM_MEDIA_ACTION_RENAME="Rename item"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_media.ini]: alpha-order

File: administrator/language/en-GB/en-GB.com_media.ini [hunk 6]
```diff
@@ -70,6 +71,7 @@ COM_MEDIA_SERVER_ERROR="Internal server error"
 COM_MEDIA_SHARE="Share the file"
 COM_MEDIA_SHARE_DESC="Share the file with others. Anyone with link can view the file"
 COM_MEDIA_SHARE_COPY_FAILED_ERROR="Error copying to clipboard"
+COM_MEDIA_TOOLBAR_LABEL="Media Manager"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_media.ini]: alpha order

File: language/en-GB/en-GB.com_media.ini [hunk 7]
```diff
@@ -3,7 +3,7 @@
 ; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
 ; Note : All ini files need to be saved as UTF-8
 
-COM_MEDIA_ACTIN_RENAME="Rename item"
+COM_MEDIA_ACTION_RENAME="Rename item"
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_media.ini]: alpha order

File: administrator/language/en-GB/en-GB.com_media.ini (hunk 5 above)
Comment (wilsonge) [reply administrator/language/en-GB/en-GB.com_media.ini]: Fixed

File: administrator/language/en-GB/en-GB.com_media.ini (hunk 6 above)
Comment (wilsonge) [reply administrator/language/en-GB/en-GB.com_media.ini]: Fixed

File: language/en-GB/en-GB.com_media.ini (hunk 7 above)
Comment (wilsonge) [reply language/en-GB/en-GB.com_media.ini]: Fixed

File: administrator/components/com_media/tmpl/media/default_texts.php (hunk 4 above)
Comment (wilsonge) [reply administrator/components/com_media/tmpl/media/default_texts.php:13]: Fixed :)

File: administrator/components/com_media/resources/scripts/components/toolbar/toolbar.vue (hunk 3 above)
Comment (wilsonge) [reply administrator/components/com_media/resources/scripts/components/toolbar/toolbar.vue]: *facepalm* thanks fixed

**Comment by wilsonge** [comment]
@brianteeman Would you be ok to the changes of an action list to type button - it is just as simple as changing the markup where you commented - but the styling is more complicated and I don't really have a huge amount of time to invest into trying to make it all work nicely

**Comment by brianteeman** [comment]
I will bow to @zwiastunsw expertise

**Comment by zwiastunsw** [comment]
1) In my opinion `role="navigation"` is redundant, but we have agreed with Brian that it must be because Joomla supports IE11 (IE11 does not support HTML5 tags). So I suggest you add it.
2) The search field has no associated label. Add it.
3) Add role="search" to `<div class="media-view-search-input">`
4) Do not change the native role of the `a` tag. If you need to use a button, use the `button`. These are action buttons, not links. See in the illustration what happens when you use `a` tag instead of a button. 
buttons

Code in file directory.vue is not correct. Change `<a href="#"...>` to `<button ...`
```html
        <a href="#" class="media-browser-select"
          @click.stop="toggleSelect()"
          :aria-label="translate('COM_MEDIA_TOGGLE_SELECT_ITEM')"
          @focus="focused(true)" @blur="focused(false)">
        </a>
        <div class="media-browser-actions" :class="{'active': showActions}">
            <a href="#" class="action-toggle" role="button"
              :aria-label="translate('COM_MEDIA_OPEN_ITEM_ACTIONS')" @keyup.enter="openActions()"
               @focus="focused(true)" @blur="focused(false)" @keyup.space="openActions()">
                <span class="image-browser-action fa fa-ellipsis-h" aria-hidden="true"
                      @click.stop="openActions()"></span>
            </a>
            <div v-if="showActions" class="media-browser-actions-list">
                <a href="#" class="action-rename" ref="actionRename" @keyup.enter="openRenameModal()"
                  :aria-label="translate('COM_MEDIA_ACTION_RENAME')"
                  @focus="focused(true)" @blur="focused(false)" @keyup.esc="showActions = false">
                    <span class="image-browser-action fa fa-text-width" aria-hidden="true"
                          @click.stop="openRenameModal()"></span>
                </a>
                <a href="#" class="action-delete" @keyup.enter="openConfirmDeleteModal()"
                  :aria-label="translate('COM_MEDIA_ACTION_DELETE')"
                   @focus="focused(true)" @blur="focused(false)" @keyup.esc="showActions = false">
                    <span class="image-browser-action fa fa-trash" aria-hidden="true" @click.stop="openConfirmDeleteModal()"></span>
                </a>
            </div>
        </div>
```

**Comment by wilsonge** [comment]
Done 1, 2 and 3. 4 is what I was discussing with Brian. The CSS is slightly beyond my ability to fix - so I'll leave that for someone else to take a stab at in a different PR

**Comment by wilsonge** [comment]
Just figured it out. Links made into buttons

**Comment by wilsonge** [comment]
The button change makes the system tests fail - I *think* https://github.com/joomla/test-system/pull/52 should fix that. Once you guys are happy with this PR i'll merge that up front

Back to the stage of I think it's only the checkboxes left

**Comment by brianteeman** [comment]
For the checkboxes see #23970

**Comment by zwiastunsw** [comment]
@wilsonge : read about links vs. button: [Links vs. Buttons in Modern Web Applications](https://marcysutton.com/links-vs-buttons-in-modern-web-applications)

**Comment by wilsonge** [comment]
So can we run a last test on this and ensure that only the checkbox issue is left to solve in a separate place?

**Comment by zwiastunsw** [comment]
@wilsonge : Which files have you changed?

**Comment by brianteeman** [comment]
I will try to find some time tomorrow to take a look. because of the npm it takes a long time to build on windows (as you saw)

**Comment by zwiastunsw** [comment]
1. When the screen reader is on, the Enter and Space keys do not activate the action buttons (Download, Rename item...) 
2. "Open item actions" button should be a [menu button](https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton)
3. The action buttons should be grouped by using a list elements (ul, li).
4. Use the arrow keys to move the focus between action items.

**Comment by wilsonge** [comment]
Done 1 and 3. I've done 2 and 4 on the image thing only. Unfortunately I need to head out but will try and find some time to do them on directory, file and video areas later today

**Comment by wilsonge** [comment]
Pushed - @zwiastunsw this should cover all the things you mentioned

**Comment by wilsonge** [comment]
@zwiastunsw @brianteeman Can you guys give this a final test - I understand there's more to do - but I'm being dragged back onto management responsibilities - handover to Harold and merging in j3 etc. So would like to get this in asap as a step in the right direction and then we can build up a list of next steps/you guys can try this yourselves (i promise it's not that bad!)

**Comment by brianteeman** [comment]
@wilsonge I will try to find some time tonight - I am working on some stuff for the marketing team thats taking my available time

**Comment by zwiastunsw** [comment]
@wilsonge:  We've made an appointment today for a Skype session

**Comment by brianteeman** [comment]
Is it perfect - no
Is it better than it was yesterday - yes

If the non graphical view was also fixed then I believe that would be a reasonable accommodation

**Comment by wilsonge** [comment]
I'm going to do something separate for the list view - I was having issues with the icons showing - I can push what I had tonight - because I don't think it was a JS issue - it seemed something CSS - that I'm sure i was just being stupid with

**Pull request merged**